### ISSUE TITLE
Handle corrupt Midi-File

### ISSUE THREAD

**Issue by FischertBilligheim** [melanchall/drywetmidi#69]
Hi Max,

the midi-file in the attached ZIP-file seems to be corrupt - an "unknown-channel-event" is thrown.
Interestingly, I created it with drywetmidi - nevertheless it cannot be loaded anymore.

I will try to findout, how I created this problem in the midi-file - this is not the topic of this issue.

My question/issue is: 
Other midi-programs can open it, it seems that they can read out the "good" parts of the midifile. 
So would it be possible to create an additional "InvalidPolicy" for that example of a corrupt-midifile?

Thanks and best regards
Thomas

[CorruptMidi.zip](https://github.com/melanchall/drywetmidi/files/4130262/CorruptMidi.zip)

Additional hint:
In Debugger, I set a breakpoint at the line where the exception is thrown. I could verify, that the next Event is correctly read by changing the instruction-pointer……..

**Comment by melanchall** [comment]
Hi Thomas,

I can create a policy to bypass such errors, but that's not a good decision. Invalid channel event is prohibited by MIDI standard. There is no guaranteed way to handle such events. In common case it means the file is corrupted. Yes, in your example you can read next event but program will fail anyway since there are another errors in the file.

Windows Media Player fails to play this file, jetAudio plays only 5 seconds. So the file is absolutely corrupted.

If this file created with DryWetMIDI and the error is caused by the library, it's a serious bug. Please find how the file was created.

**Comment by FischertBilligheim** [comment]
Hi Max, 
thanks a lot again for your very fast response!

I thought this is the only corrupt part in the file, because I did not get another exception in the file - and i could play it afterwards 
And - as I mentioned - I could read the corrupt file e.g. with my DAW Studio-one - and not only with that.

In the features list of drywetmidi in the ReadMe  is explicitly noted, that midi-file errors can be handled, so I proposed to extend the functionality with the results of my corrupt file.

Sure - will try to findout the reason - but I will need some time to do that.

I don‘t think it‘s an issue of the library. I assume I manipulated data on a low level and in parallel with the notes-manager.
Shall I close this issue and create a new one if I find a bug in drywetmidi?

**Comment by melanchall** [comment]
Thomas,

If you find a bug in the library, please create new issue. BUT please don't close **this** issue, I'll implement new policy within it.

**Comment by melanchall** [comment]
I've added `UnknownChannelEventPolicy` to `ReadingSettings`. For your file you should use this settings:

```csharp
var midiFile = MidiFile.Read("Another brick in the wall (Pink Floyd).MID", new ReadingSettings
{
    UnknownChannelEventPolicy = UnknownChannelEventPolicy.SkipStatusByte,
    InvalidChannelEventParameterValuePolicy = InvalidChannelEventParameterValuePolicy.SnapToLimits
});
```

I've added following policies:

* `Abort` – throw exception
* `SkipStatusByte` – skip only invalid status byte
* `SkipStatusByteAndOneDataByte` – skip status byte and one byte more
* `SkipStatusByteAndTwoDataBytes` – skip status byte and two bytes more
* `UseCallback` – manually decide whether skip event or abort reading using `ReadingSettings.UnknownChannelEventCallback`

Please test the new API.

**Comment by FischertBilligheim** [comment]
Wow - again your speed and reaction time is unbelievable....👍👍👍👍👍
I surely will test it.....
Thanks really a lot, Max!

**Comment by FischertBilligheim** [comment]
Hi Max,
I today tested your last changes: As expected, I now can load the corrupt file, and after a save the corruption is corrected. Thanks a lot for your great work!

I then tried to found out, what happened to get such a corrupt file. 
This is really difficult to find out. I think I used a specific midi-sequencer - "PowerTracksPro" and additionally  my own C#-program, based on DryWetMIDI, to edit the midi-file. But until now, I could not reproduce that issue.

So I think I will close this issue - and in the future I nevertheless will try to find out what the reason for the corrupt file was.

Thanks again a lot for these new policies...
